lib/oxigraph/src/sparql/dataset.rs

```diff
@@ -1,18 +1,235 @@
+use crate::model::NamedNodeRef;
 use crate::sparql::QueryDataset;
 use crate::storage::numeric_encoder::{
     insert_term, Decoder, EncodedTerm, EncodedTriple, StrHash, StrHashHasher, StrLookup,
 };
 use crate::storage::{CorruptionError, StorageError, StorageReader};
+use hdt::Hdt;
 use oxrdf::Term;
 use oxsdatatypes::Boolean;
 use spareval::{ExpressionTerm, ExpressionTriple, InternalQuad, QueryableDataset};
 use std::cell::RefCell;
 use std::collections::hash_map::Entry;
 use std::collections::HashMap;
 use std::hash::BuildHasherDefault;
+use std::io::{Error, ErrorKind};
 use std::iter::empty;
+use std::str::FromStr;
 use std::sync::Arc;
 
+pub struct HDTDataset {
+    /// Path to the HDT file.
+    path: String,
+
+    /// HDT interface.
+    hdt: Arc<Hdt>,
+}
+
+/// Boundry over a Header-Dictionary-Triplies (HDT) storage layer.
+pub struct HDTDatasetView {
+    // collection of HDT files in the dataset
+    hdts: Vec<HDTDataset>,
+}
+
+/// Cloning opens the same file again.
+impl Clone for HDTDatasetView {
+    fn clone(&self) -> HDTDatasetView {
+        let mut hdts: Vec<HDTDataset> = Vec::new();
+        for dataset in &self.hdts {
+            hdts.push(HDTDataset {
+                path: dataset.path.clone(),
+                hdt: Arc::<Hdt>::clone(&dataset.hdt),
+            })
+        }
+
+        Self { hdts }
+    }
+}
+
+impl HDTDatasetView {
+    pub fn new(paths: &[String]) -> Self {
+        let mut hdts: Vec<HDTDataset> = Vec::new();
+        for path in paths {
+            // TODO catch error and proceed to next file?
+            let hdt = Hdt::new_from_path(std::path::Path::new(&path)).unwrap();
+            hdts.push(HDTDataset {
+                path: path.to_string(),
+                hdt: Arc::new(hdt),
+            })
+        }
+
+        Self { hdts }
+    }
+}
+
+/// Create the correct OxRDF term for a given resource string.  Slow,
+/// use the appropriate method if you know which type (Literal, URI,
+/// or blank node) the string has. Based on
+/// https://github.com/KonradHoeffner/hdt/blob/871db777db3220dc4874af022287975b31d72d3a/src/hdt_graph.rs#L64
+fn hdt_bgp_str_to_term(s: &str) -> Result<Term, Error> {
+    match s.chars().next() {
+        None => Err(Error::new(ErrorKind::InvalidData, "empty input")),
+
+        // Double-quote delimters are used around the string.
+        Some('"') => match s.rfind('"') {
+            None => Err(Error::new(
+                ErrorKind::InvalidData,
+                format!("missing right quotation mark in literal string {s}"),
+            )),
+
+            Some(_) => match Term::from_str(s) {
+                Ok(s) => Ok(s),
+                Err(e) => Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("literal parse error {e} for {s}"),
+                )),
+            },
+        },
+
+        // Underscore prefix indicating an Blank Node.
+        Some('_') => {
+            let term = match oxrdf::BlankNode::new(&s[2..]) {
```

Comment:
An other option is `BlankNode::from_str(s)` if you prefer

Reply:
I'll have to test this out, perhaps it is no longer applicable with the new implementation.  But we hit some problems early on with the conversion and there was an issue when including the leading "_" chars.

Reply:
Weird, `BlankNode::from_str` should work with `_:ad233` like identifiers